substrate/client/network/sync/src/pending_responses.rs

```diff
@@ -0,0 +1,108 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! [`PendingResponses`] is responsible for keeping track of pending responses and
+//! polling them.
+
+use futures::{
+	channel::oneshot,
+	future::BoxFuture,
+	stream::{BoxStream, Stream},
+	FutureExt, StreamExt,
+};
+use libp2p::PeerId;
+use log::error;
+use sc_network::request_responses::RequestFailure;
+use sc_network_common::sync::PeerRequest;
+use sp_runtime::traits::Block as BlockT;
+use std::task::{Context, Poll};
+use tokio_stream::StreamMap;
+
+/// Response result.
+type ResponseResult = Result<Result<Vec<u8>, RequestFailure>, oneshot::Canceled>;
+
+/// A future yielding [`ResponseResult`].
+type ResponseFuture = BoxFuture<'static, ResponseResult>;
+
+/// An event we receive once a pending response future resolves.
+pub(crate) struct ResponseEvent<B: BlockT> {
+	pub peer_id: PeerId,
+	pub request: PeerRequest<B>,
+	pub response: ResponseResult,
+}
+
+/// Stream taking care of polling pending responses.
+pub(crate) struct PendingResponses<B: BlockT> {
+	/// Pending responses
+	pending_responses: StreamMap<PeerId, BoxStream<'static, (PeerRequest<B>, ResponseResult)>>,
+}
+
+impl<B: BlockT> PendingResponses<B> {
+	pub fn new() -> Self {
+		Self { pending_responses: StreamMap::new() }
+	}
+
+	pub fn insert(
+		&mut self,
+		peer_id: PeerId,
+		request: PeerRequest<B>,
+		response_future: ResponseFuture,
+	) {
+		let request_type = request.get_type();
+
+		if self
+			.pending_responses
+			.insert(
+				peer_id,
+				Box::pin(async move { (request, response_future.await) }.into_stream()),
+			)
+			.is_some()
+		{
+			error!(
+				target: crate::LOG_TARGET,
+				"Discarded pending response from peer {peer_id}, {request_type:?}.",
```

Comment:
After the refactoring, this error is triggered. This was not the case before. This is likely the reason why tests in `sc-network-test` are failing. Looking into it.

Reply:
Fixed. We had to manually remove resolved responses from `StreamMap`, otherwise they were overwritten before the underlying stream had a chance to yield `None`.